doc/meson.build

```diff
@@ -0,0 +1,128 @@
+                                   input: ['geany.css', 'geany.txt'],
+                                   output: ['geany.html'],
+                                   build_by_default: true,
+                                   command: cmd)
+        custom_target('hacking.html',
+                      input: ['geany.css', '../HACKING'],
+                      output: ['hacking.html'],
+                      build_by_default: true,
+                      command: cmd)
+    endif
+    # geany.html is installed as index.html
+    custom_target('index.html',
+                  input: [geany_html],
+                  output: ['index.html'],
+                  command: [ln, '-f', '@INPUT@', '@OUTPUT@'],
+                  install: true,
+                  install_dir: join_paths(cdata.get('GEANY_DOC_DIR'), 'html'))
+endif
+
+install_data('geany.txt',
+             rename: 'manual.txt',
+             install_dir: cdata.get('GEANY_DOC_DIR'))
+install_data('../scintilla/License.txt',
+             rename: 'ScintillaLicense.txt',
+             install_dir: cdata.get('GEANY_DATA_DIR'))
+install_data('../scintilla/lexilla/License.txt',
+             rename: 'LexillaLicense.txt',
+             install_dir: cdata.get('GEANY_DATA_DIR'))
+
+# Normally, doxygen is required for the gtkdoc headers
+# but it can be disabled if really neded (e.g. if plugins are also disabled),
+# packagers must not disable this!
+doxygen = find_program('doxygen', required: false)
+if doxygen.found()
+    doxcfg = configuration_data()
+    doxcfg.merge_from(cdata)
+    doxcfg.set('GIRONLY', '@internal')
+    doxcfg.set('HTML', 'YES')
+    doxcfg.set('XML', 'NO')
+    doxcfg.set('SORT', 'YES')
+
+    doxgicfg = doxcfg
+    doxgicfg.set('GIRONLY', '')
+    doxgicfg.set('HTML', 'NO')
+    doxgicfg.set('XML', 'YES')
+    doxgicfg.set('SORT', 'NO')
+
+    dox = configure_file(input: 'Doxyfile.in',
+                         output: 'Doxyfile',
+                         configuration: doxcfg)
+    doxgi = configure_file(input: 'Doxyfile.in',
+                           output: 'Doxyfile-gi',
+                           configuration: doxgicfg)
+
+    dep_doxygen = declare_dependency(sources: ['plugins.dox',
+                                               'pluginsignals.c',
+                                               'pluginsymbols.c',
+                                               'stash-example.c',
+                                               'stash-gui-example.c'])
+
+    custom_target('doxygen.stamp',
+                  input: dox,
+                  output: ['doxygen.stamp'],
+                  command: [doxygen, '@INPUT@', '&&', 'touch', '@OUTPUT@'],
+                  depends: libgeany,
+                  build_by_default: true)
+    doxy_xml = custom_target('doxygen-gi.stamp',
+                             input: doxgi,
+                             output: ['doxygen-gi.stamp'],
+                             command: [doxygen, '@INPUT@', '&&', 'touch', '@OUTPUT@'],
```

Comment:
Meson's command kwarg does not accept shell commands, only execve() commands. If you need the former, use `['sh', '-c', '...']` or a python script.

Reply:
Thinking about it, I think I don't actually need the stamp file as the `custom_target` object can be used as a dependency itself. Inherited from Makefile.am where the stamp file ensures proper ordering.

Reply:
Grr, https://github.com/mesonbuild/meson/issues/2320

Reply:
Does it output to its own directory exclusively created by doxygen?

Because in another project of mine we have a doxygen custom_target which defines `output: 'man3'`and that works as far as meson/ninja is concerned. Even though the output is a directory node rather than a file node, it is an output without "/" in the name, and ninja can recognize when a directory is "up to date".

Doxygen... so fun.

Reply:
If directories as output works then we're green, I'll try

Reply:
No, does not work. The gtkdoc header is not regenerated if doxygen re-runs to generate $builddir/xml/*.xml. The timestamp of the xml directory isn't updated by the re-run of doxygen.

Reply:
Also, meson can be a PITA. It seems that I cannot use a `declare_dependency()` object in the `depends:` kwarg of `custom_target()`

Reply:
Ah yes... we don't then depend on those man3 pages, we just install them directly.

In that case, your best bet would seem to be a python script `['foo.py', '@OUTPUT@', doxygen, '@INPUT@']` that runs `sys.argv[2:]` and then updates the mtime of the stamp file.

Reply:
Also regarding the declare_dependency, a declare_dependency is a way of producing a custom defined pkg-config style `dependency()` object, hence why it can also expose c_args and link_args, which have no meaning to a custom_target.

You can just list the files directly as custom_target `depends: `

Reply:
> You can just list the files directly as custom_target depends: 

No, I can't :(
`depends: [files(…), libgeany],`
-> `ERROR: custom_target keyword argument 'depends' was of type array[File | SharedLibrary] but should have been array[BuildTarget | CustomTarget]`


I'm also unable to specify the extra C files as input
`input: [doxyfile_gi] + files(…),`
-> `[2/14] /usr/bin/sh -c 'doxygen @INPUT0@ && touch doc/doxygen.stamp'`

(INPUT0 should just use the first input file)

Reply:
Apologies, I meant `depend_files: `

Reply:
OK. `depend_files` seems to work, but not my shell wrapper: `@OUTPUT@` is not substituted when `command: [sh, '-c', 'doxygen @INPUT@ && touch @OUTPUT@'],` is run: `/usr/bin/sh -c 'doxygen doc/Doxyfile && touch @OUTPUT@'`

Reply:
Please do not pass arguments by embedding them inside a shell script. There is no `shellquote('@OUTPUT@')`, you cannot guarantee it has a shell-safe name like that.

Instead, use
```meson
command: [sh, '-c', 'doxygen "$1" && touch "$2"', '_argv0', '@INPUT@', '@OUTPUT@'],
```

Which will safely pass the filenames as command line *arguments*, and allow the shell to reference those filenames using variables, since variables can be safely quoted regardless of their content.

This *also* means that meson has a much more accurate understanding of how to replace those strings.

Specifically, you cannot replace both INPUT and OUTPUT in the *same* string. https://github.com/mesonbuild/meson/issues/1864

Reply:
Thanks, your shell magic works.

This all shows me why I don't actually like meson.

Reply:
As a matter of curiosity, why is this the fault of meson rather than the fault of embedding shell scripts as data?

```meson
command: [sh, '-c', 'doxygen @INPUT@ && touch @OUTPUT@'],
```

is inherently broken, because what happens if `@INPUT@` or `@OUTPUT@` have spaces in them? Or any of a number of other character sequences that would break a shell script?

Reply:
Then let it be `command: [sh, '-c', 'doxygen "@INPUT@" && touch "@OUTPUT@"'],`

Anyway, the root problem is that I can't specify output files in subdirectories, this is the only reason for the stamp file (and shell wrapper) at all. And the discussion on  mesonbuild/meson#2320 doesn't look like meson cares much about users.

Reply:
The `"` character is one of the "would break a shell script" characters I mentioned.

The root problem of specifying output files in subdirectories is a legitimate point, which nevertheless (fortunately?) does not affect the majority of use cases. For unrelated reasons, I actually do want to add a doxygen function to meson, which *would* be able to do behind the scenes stuff and track output files in subdirectories...